ARMmbed/mbedtls
# Restrict usage of WANT_READ to support event-driven IO

__Problem:__ The error code `MBEDTLS_ERR_SSL_WANT_READ` is overloaded with two different meanings:

1.  Firstly, it can indicate that there is no further data available from the underlying transport

2. Secondly, it can indicate that further processing of internal buffers is necessary and that the previously called function should be called again.

The lack of separation between these two meanings impedes the use of event-driven IO, because the user cannot know when to idle on the underlying transport and when to immediately call the previous function again. This was also the problem that triggered the issue in the first place in #772.

__This PR:__ Restricts the meaning of `MBEDTLS_SSL_WANT_READ` to case (1), looping in the cases where it was previously used for the purpose of (2). 

Further, it provides the following enhancements to using and testing event-driven I/O:

__Test infrastructure:__ 
* A `pack=TIME` option for our UDP proxy allowing to merge multiple records into a single UDP datagram.
* An 'UDP proxy wrapper' which silently puts a server binary like `ssl_server2` behind a UDP proxy. This is useful e.g. to run our test-suites with UDP packing added everywhere without the need to change server application or test script.
* Add `event=1` option to `ssl_server2` and `ssl_client2` to enable event-driven IO. With that, the applications idle when receiving a `WANT_READ` or `WANT_WRITE`, until they either receive an event from the timer or the underlying transport.

__Library:__
* Add a new public polling function for `mbedtls_net_context_t` allowing to wait for events on mbed TLS provided network contexts.
* Add a new public function `mbedtls_ssl_check_pending` to check whether the internal buffers still contain data to be processed. This is an inevitable addition because a user employing event driven IO needs that information to decide when he/she can idle on the underlying transport. Note that this is different from `mbedtls_ssl_get_bytes_avail` which only concerns application data. A simple usecase demonstrating the need for `mbedtls_ssl_check_pending` is the case when multiple application data records are contained in a single UDP datagram. Having fully read the data from the first datagram, `mbedtls_ssl_get_bytes_avail` will return `0`, but idling on the underlying transport might deadlock.

__References:__ Originally triggered by #772, internal reference IOTSSL-1204

@mpg Please review.
